### ISSUE TITLE
MappedDistribution

### ISSUE THREAD

**Issue by mohamed82008** [TuringLang/DistributionsAD.jl#12]
Sometimes it is useful to be able to define a multivariate distribution on iid variables by generating distributions on the fly which use different distribution parameters each variable according to a certain rule/function. Defining an efficient `logpdf` and adjoint can give significant computational savings. This is similar to the Soss `For` combinator.

**Comment by cscherrer** [comment]
Here's my current setup:
```julia
struct For{F,T,D,X} 
    f :: F  
    θ :: T
end
```

where...

- `F` and `T` are specified in the struct
- `D` is the distribution returned
- `X` is the `eltype` of that distribution (unfortunately, often not available directly from `D`)

Some example use cases:
```julia
# T = NTuple{2,Int}
x ~ For(10,3) do i,j
    Bernoulli(j/i)
end
```

```julia
# T = Base.Generator{Base.OneTo{Int64},Base.var"#174#175"{Array{Float64,2}}}
y ~ For(eachrow(X)) do xrow
    Normal(xrow' * β, 1)
end
```

We'll have different methods for `rand`, `logpdf`, etc based mostly on `T`.

Also, I currently have the following restrictions:
1. `D` is consistent across indices
2. `support(d::D)` is consistent across indices

Currently this targets "array-like" results, but in principle `T` can be anything, for example an iterator or `Real` (for function spaces, GPs, etc).

**Comment by mohamed82008** [comment]
I don't think we need a restriction on `D` being the same. The `logpdf` can be something like this:
```julia
function logpdf(dist::For, x::AbstractArray)
    @assert size(dist.θ) == size(x)
    return sum(1:length(d.θ)) do i
        logpdf(dist.f(d.θ[i]), x[i])
    end
end
rand(dist::For) = rand.(dist.f.(dist.θ))
```
Whether `f` returns the same distribution or not, this should be inferrable by the Julia compiler.

**Comment by mohamed82008** [comment]
```julia
eltype(dist::For) = mapreduce(i -> eltype(dist.f(dist.θ[i])), promote_type, 1:length(dist.θ))
```

**Comment by mohamed82008** [comment]
Note that the above is a dynamically sized distribution. We can also get free specialization and inlining for small, fixed-size distributions when using `θ::StaticArray`.

**Comment by mohamed82008** [comment]
I think for Tracker
```julia
sum(logpdf.(dist.f.(dist.θ), x))
```
will be faster than
```julia
sum(1:length(dist.θ)) do i
    logpdf(dist.f(dist.θ[i]), x[i])
end
```
So if either `θ` or `x` is a `TrackedArray`, all intermediates will also be `TrackedArray`s not `TrackedReal`s.

**Comment by cscherrer** [comment]
> I don't think we need a restriction on D being the same.
The vast majority of models will satisfy this anyway, and it often opens up opportunities for optimization. For example, in cases where `d.f` maps to continuous distributions, how can we determine the bijection to ℝⁿ? Parameterizing by `D` makes this trivial.

One thing I've found a bit tricky is make useful type information available without much computational cost. Unfortunately in Julia, we can't just ask a function about its codomain, so instantiating a `For` requires some computation in order to determine the types. To this point, I've been trying to make construction cheap by assuming distributions and supports are consistent, and just computing them for a single index at construction time. Your `eltype` suggestion,
```julia
eltype(dist::For) = mapreduce(i -> eltype(dist.f(dist.θ[i])), promote_type, 1:length(dist.θ))
```
is appealing, but would require O(n) instantiation cost.

Above I suggested `For` might also be useful for building distributions over function spaces. I may disagree with myself on this point, because it drops the conditional independence assumption of other `For` instances, and would require adding some way to specify covariance.

Finally, we had some [recent discussion on Discourse](https://discourse.julialang.org/t/parallel-reductions/30180/8) about the best approach for parallelism, which will be important for many cases.

**Comment by cscherrer** [comment]
Cleaning this up a bit in Soss, here's the current state:
https://github.com/cscherrer/Soss.jl/blob/dev/src/for.jl
Should be able to get a PR submitted today.

There's also `iid`, which is like `For` but without the distributional dependence on indices. I have a curried form, which I usually use like this:

```julia
x ~ Normal() |> iid(N)
```

**Comment by mohamed82008** [comment]
Thanks for the PR @cscherrer and sorry for the late review; I was busy the last few weeks. I will review your PR asap.
